### PR TITLE
Implemented upb_enumvaldef, for storing information about enumvals.

### PR DESCRIPTION
**Pull request by haberman**
Previously `upb_enumdef` (the upb equivalent of `proto2::EnumDescriptor`) just contained `name -> number` and `number -> name` maps. There was no explicit def for each individual enum value.

We need `upb_enumvaldef` for two reasons:

1. (immediate need): we need to be able to look up all enum values underneath a given message. Thanks to proto's weird scoping rules, these enum values are *not* in the scope of the enum itself, so the `upb_enumdef`'s table cannot be used to perform this lookup. We unfortunately have to look up the enum values in the `upb_symtab` global map, since the value could come from any number of enums under the message.
2. (future need): Any enum value can have custom options attached, and we need a place to store these custom options.

There is minor code size growth, which is fine, but unfortunately this regresses the Ads benchmark for schema loading by 10%.

```
name                                      old time/op  new time/op  delta                                                                                                     
ArenaOneAlloc                             20.9ns ± 0%  21.4ns ± 1%   +2.25%  (p=0.000 n=12+12)           
ArenaInitialBlockOneAlloc                 6.29ns ± 0%  6.28ns ± 0%   -0.05%  (p=0.000 n=12+11)           
LoadDescriptor_Upb                        53.6µs ± 1%  54.8µs ± 1%   +2.24%  (p=0.000 n=12+12)           
LoadAdsDescriptor_Upb                     3.13ms ± 0%  3.45ms ± 0%  +10.22%  (p=0.000 n=12+11)           
LoadDescriptor_Proto2                      240µs ± 0%   239µs ± 0%   -0.48%  (p=0.000 n=12+12)           
LoadAdsDescriptor_Proto2                  14.5ms ± 0%  14.4ms ± 0%   -0.63%  (p=0.000 n=11+10)           
Parse_Upb_FileDesc<UseArena,Copy>         10.8µs ± 0%  11.1µs ± 0%   +3.24%  (p=0.000 n=12+11)           
Parse_Upb_FileDesc<UseArena,Alias>        9.00µs ± 0%  9.35µs ± 0%   +3.87%  (p=0.000 n=12+12)                                                                                
Parse_Upb_FileDesc<InitBlock,Copy>        10.3µs ± 0%  10.7µs ± 0%   +3.39%  (p=0.000 n=12+12)           
Parse_Upb_FileDesc<InitBlock,Alias>       8.67µs ± 0%  8.87µs ± 0%   +2.29%  (p=0.000 n=12+12)           
Parse_Proto2<FileDesc,NoArena,Copy>       29.3µs ± 0%  29.5µs ± 0%   +0.43%  (p=0.000 n=12+11)           
Parse_Proto2<FileDesc,UseArena,Copy>      20.4µs ± 2%  20.5µs ± 2%     ~     (p=0.319 n=12+12)           
Parse_Proto2<FileDesc,InitBlock,Copy>     16.7µs ± 0%  16.7µs ± 0%   -0.47%  (p=0.000 n=10+10)           
Parse_Proto2<FileDescSV,InitBlock,Alias>  16.8µs ± 0%  17.5µs ± 1%   +4.13%  (p=0.000 n=12+12)           
SerializeDescriptor_Proto2                5.31µs ± 1%  5.34µs ± 1%   +0.56%  (p=0.024 n=12+12)           
SerializeDescriptor_Upb                   12.2µs ± 0%  12.3µs ± 0%   +0.90%  (p=0.000 n=12+11)           
                                                                                                                                                                              
                                                                                                                                                                              
    FILE SIZE        VM SIZE                                                                                                                                                  
 --------------  --------------                                                                                                                                               
  +3.7%    +669  +3.6%    +568    upb/def.c                                                                                                                                   
     +43%    +416   +45%    +416    create_enumdef                                                                                                                            
    [NEW]    +148  [NEW]    +104    upb_enumvaldef_name                                                                                                                       
    [NEW]    +135  [NEW]     +88    upb_enumdef_lookupname                                                                                                                    
    [NEW]    +134  [NEW]     +88    upb_enumdef_lookupnum                                                                                                                     
    [NEW]    +102  [NEW]     +56    upb_enumvaldef_number                                                                                                                     
    +1.0%     +48  +1.1%     +48    _upb_symtab_addfile                                                                                                                       
    +0.6%     +16  +0.6%     +16    create_msgdef                                                                                                                             
    +1.1%     +16  +1.2%     +16    resolve_fielddef                                                                                                                          
    -7.6%      -8 -12.5%      -8    field_number_cmp                                                                                                                          
    -3.2%     -64  -3.3%     -64    create_fielddef                                                                                                                           
    [DEL]    -129  [DEL]     -88    upb_enumdef_iton                                                                                                                          
    [DEL]    -145  [DEL]    -104    upb_enumdef_ntoi                                                                                                                          
  +6.5%     +39  [ = ]       0    [section .strtab]                                                                                                                           
  +0.3%     +16  +0.3%     +16    upb/text_encode.c                                                                                                                           
    +1.7%     +16  +1.8%     +16    txtenc_field                                                                                                                              
  +0.6%     +15  [ = ]       0    upb/msg.c                                                                                                                                   
    +4.9%     +15  [ = ]       0    _upb_array_realloc                                                                                                                        
  -0.0%      -4  [ = ]       0    upb/table.c                                                                                                                                 
    +1.2%     +14  [ = ]       0    upb_inttable_insert                                                                                                                       
    +7.5%      +9  [ = ]       0    upb_strtable_iter_value                                                                                                                   
    +2.9%      +8  [ = ]       0    upb_strtable_lookup2                                                                                                                      
    +4.5%      +8  [ = ]       0    upb_strtable_next                                                                                                                         
    -0.2%      -1  [ = ]       0    upb_strtable_remove                                                                                                                       
    -0.2%      -7  [ = ]       0    upb_inttable_compact                                                                                                                      
    -5.8%      -7  [ = ]       0    upb_inttable_init                                                                                                                         
    -2.9%     -13  [ = ]       0    upb_inttable_sizedinit                                                                                                                    
    -5.6%     -15  [ = ]       0    upb_inttable_begin                                                                                                                        
  -0.5%     -15  [ = ]       0    tests/conformance_upb.c                                                                                                                     
    -6.8%     -15  [ = ]       0    main                                                                                                                                      
  -0.1%     -16  -0.1%     -16    upb/json_decode.c                                                                                                                           
    -0.8%     -16  -0.8%     -16    jsondec_value                                                                                                                             
  -0.5%     -32  -0.5%     -32    [section .rodata]                                                                                                                           
   -26.1%      -6 -26.1%      -6    parse_proto.msg                                                                                                                           
    -0.4%     -26  -0.4%     -26    [section .rodata]                                                                                                                         
  -0.5%     -35  [ = ]       0    bazel-out/k8-opt/bin/external/com_google_protobuf/google/protobuf/descriptor.upb.c
    -5.7%      -6  [ = ]       0    google_protobuf_EnumDescriptorProto_msginit                                                                                               
    -6.4%      -7  [ = ]       0    google_protobuf_DescriptorProto_ExtensionRange_msginit               
    -6.0%      -7  [ = ]       0    google_protobuf_EnumDescriptorProto_EnumReservedRange_msginit                                                                             
    -7.7%      -7  [ = ]       0    google_protobuf_EnumOptions_msginit                                                                                                       
    -7.6%      -8  [ = ]       0    google_protobuf_EnumValueDescriptorProto_msginit                                                                                          
 -17.0%    -533  [ = ]       0    [Unmapped]                                                                                                                                  
  +0.1%    +104  +0.4%    +536    TOTAL
```

Unfortunately I don't see an easy way around this. We're forced into the extra table entries by proto's weird enum scoping rules.

### CODE REVIEW

File: upb/def.c [hunk 1]
```diff
@@ -270,10 +279,31 @@ const upb_filedef *upb_enumdef_file(const upb_enumdef *e) {
...
+}
+
+const upb_enumvaldef *upb_enumdef_lookupnum(const upb_enumdef *def, int32_t num) {
+  upb_value v;
+  return upb_inttable_lookup(&def->iton, num, &v) ? upb_value_getconstptr(v)
+                                                  : NULL;
+}
+
+const upb_enumvaldef *upb_enumdef_value(const upb_enumdef *e, int i) {
+  UPB_ASSERT(i >= 0 && i < e->value_count);
```
Comment (fowles) [inline upb/def.c:301]: if you write this as `min <= i && i < max` it reads much closer to the match notation people are used to

File: upb/def.c [hunk 2]
```diff
@@ -1727,34 +1773,35 @@ static void create_enumdef(
...
   }
 
   for (i = 0; i < n; i++) {
-    const google_protobuf_EnumValueDescriptorProto *value = values[i];
-    upb_strview name = google_protobuf_EnumValueDescriptorProto_name(value);
-    char *name2 = strviewdup(ctx, name);
-    int32_t num = google_protobuf_EnumValueDescriptorProto_number(value);
-    upb_value v = upb_value_int32(num);
+    const google_protobuf_EnumValueDescriptorProto *val_proto = values[i];
+    upb_enumvaldef *val = (upb_enumvaldef*)&e->values[e->value_count++];
```
Comment (fowles) [inline upb/def.c:1786]: I think `e->value_count` is  `i` throughout this loop.  If so, I would find it slightly easier to read if you use i consistently and then outside the loop assign `e->value_count = n`

File: upb/def.hpp [hunk 3]
```diff
@@ -335,15 +348,15 @@ class EnumDefPtr {
...
+  const EnumValDefPtr FindValueByName(const char* name) const {
+    return EnumValDefPtr(upb_enumdef_lookupnamez(ptr_, name));
   }
 
   // Finds the name corresponding to the given number, or NULL if none was
   // found.  If more than one name corresponds to this number, returns the
   // first one that was added.
-  const char* FindValueByNumber(int32_t num) const {
-    return upb_enumdef_iton(ptr_, num);
+  const EnumValDefPtr FindValueByNumber(int32_t num) const {
```
Comment (fowles) [inline upb/def.hpp:358]: const on a return value has no effect.

File: upb/def.c (hunk 2 above)
Comment (haberman) [reply upb/def.c:1786]: Done.

File: upb/def.c (hunk 1 above)
Comment (haberman) [reply upb/def.c:301]: Done.

File: upb/def.hpp (hunk 3 above)
Comment (haberman) [reply upb/def.hpp:358]: Done.

**Review by fowles — APPROVED** [review]

**Pull request merged**